### PR TITLE
Algolia search

### PR DESCRIPTION
**Pull request by wackerow**
## Description
- Adds `algoliasearch` and `@docsearch/react` packages
- Uses the `DocSearch` component from `@docsearch/react` to render a search button and modal, along with `@docsearch/css` stylesheet.
- Using this library component enables a well designed, a11y-friendly search modal out-of-box; we just need to pass it the app ID, API key and index name from Algolia (stored as env vars)
- Leaves the "Search by Algolia" logo, required by Algolio to be eligible for free open source indexing plan.
- Created `search.css` to select for docsearch elements by class, and override theming to match site theme with latest updated design.
- Removes old Search component

## Preview link:
https://deploy-preview-59--melodious-puffpuff-8e1109.netlify.app/

## Todo
- [x] Get site crawled by Algolia
- [x] Obtain app ID and API key from Algolia
- [x] Add to .env vars and remove hard coded testing keys
- [x] Check to make sure modal closes appropriately on mobile when a link is tapped. This may not work in current form.
- [x] Test everything

## Screenshots

image
image
image
image
image
image
image

### CODE REVIEW

**Comment by nloureiro** [comment]
looks good!!!!

for now I would change the hover style
https://user-images.githubusercontent.com/1120748/204748944-3c6039f4-58d1-435e-a64b-a8277b0b72a7.mov


@wackerow can we make the hover behaviour like the other buttons?

**Comment by samajammin** [comment]
@wackerow what's the status on this PR?

**Comment by wackerow** [comment]
Update: We got in contact with Algolia and we now have a working index with credentials 🎉. These credentials are accessed via environment variables, and these have been loaded into Netlify.

Remaining issue: The index is currently of the melodious-puffpuff `master` Netlify preview, so all the search results link to this domain, not geth.ethereum.org (which makes sense right now, since we have yet to migrate this codebase).

One possible path forward:
1. Pull this in now
1. Migrate site to `go-ethereum` repo under new `website` branch
1. Get Netlify deploys connected, and get a new domain name for our latest preview
1. Re-index the site* with latest preview domain while site is undergoing final reviews from geth team
1. Switch DNS to geth.ethereum.org, and re-index the site again* with this domain. 

\* The site will be automatically re-indexed every 24-hours. I've reached out to Algolia support regarding how to update the domain being crawled once we make this transition.

With this path we may have up to 24 hours where the search points people to the staging site preview. _As an alternative_, we could migrate the site first with the duckduckgo patch, update the domain being crawled on the Algolia side and wait 24 hours, _then_ merge this in. _Using this path, we may still have issues where duckduckgo has not yet recrawled the site paths either._

I would advocate for the first path myself.

- [ ] Other issue: Once we can preview this with locally references search results, we need to just test/check that search modal closes appropriately, especially on mobile, when a search result is executed.

**Comment by wackerow** [comment]
```
✨  Done in 22.35s.
```
Building fine for me locally, rerunning Netlify build

**Comment by wackerow** [comment]
Not ready yet... noticing relative links are stacking weird here... Going to `/docs/fundamentals` then searching for "Fundamentals" and following that link goes to `/docs/docs/fundamentals` and breaks. Will investigate.

File: src/components/UI/Search.tsx [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  const appId = process.env.NEXT_PUBLIC_ALGOLIA_APP_ID || '';
+  const apiKey = process.env.NEXT_PUBLIC_ALGOLIA_SEARCH_API_KEY || '';
+  const indexName = process.env.NEXT_PUBLIC_ALGOLIA_BASE_SEARCH_INDEX_NAME || '';
+
+  // TODO: Replace Algolia test keys with above env vars when ready
+  return (
+    <DocSearch appId={appId} apiKey={apiKey} indexName={indexName} transformItems={
+      items => items.map(item => ({
+        ...item,
+        url: item.url.replace('https://melodious-puffpuff-8e1109.netlify.app/', ''),
```
Comment (samajammin) [inline src/components/UI/Search.tsx:15]:
```suggestion
        url: item.url.replace('https://melodious-puffpuff-8e1109.netlify.app/, ''),
```
I suspect we want to keep the trailing slash

Comment (wackerow) [reply src/components/UI/Search.tsx:15]:
```suggestion
        url: item.url.replace('https://melodious-puffpuff-8e1109.netlify.app', ''),
```

Comment (samajammin) [reply src/components/UI/Search.tsx]: (that's what we do here: https://github.com/ethereum/ethereum-org-website/blob/dev/src/components/Search/index.tsx#L123)

Comment (wackerow) [reply src/components/UI/Search.tsx]: Hm, good call, let's see how that does.

Comment (wackerow) [reply src/components/UI/Search.tsx]: Hm, still seeing the same bug where the routes are stacking after removing the trailing slash.

Comment (wackerow) [reply src/components/UI/Search.tsx]: Actually, I may have spoke too soon. I think I was having issues with cookies/localStorage not updating when I made this change and tested, and it was using the old links. I switched to a private browser window and am having better luck. 🤞 Hopefully the Netlify preview will pass this time and we can try it out on that.

**Comment by wackerow** [comment]
I extracted the index domain to first check for an env variable, then fallback to the current Netlify preview (melodious-puffpuff), this was we can quickly update and override as needed from the Netlify backend. 

I am wondering though if instead of search for a specific domain string, we just match`https://[anything]/` and remove that instead, so all search results will have whatever domain is in there stripped out.

File: src/components/UI/Search.tsx (hunk 1 above)
Comment (wackerow) [reply src/components/UI/Search.tsx]: I think that fixed the issues of the URLs stacking, thanks!

**Comment by corwintines** [comment]
I rebuilt on Netlify and this is working. Won't add the green check when I kick off a build in Netlify for some reason. We are getting rate limited by azure on these PRs.

**Comment by wackerow** [comment]
Okay thanks... 

Related to this PR, added another PR (#163) to add a sitemap and robots.txt, similar to the blog, as this was recommended by DocSearch to improve crawling.

**Review by corwintines — APPROVED** [review]

File: src/components/UI/Header.tsx [hunk 2]
```diff
@@ -2,8 +2,9 @@ import { FC } from 'react';
 import { Box, Flex, Link, Stack, Text, useColorMode } from '@chakra-ui/react';
 import NextLink from 'next/link';
 
+import { Search } from '../UI';
```
Comment (nhsz) [inline src/components/UI/Header.tsx:5]: 👍👍👍

**Review by nhsz — APPROVED** [review]
Great job @wackerow!!!, we finally have the searchooor 🔥🔥🔥

**Pull request merged**